src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+          << std::dec << " with backtrace: " << cpp_strerror(r) << dendl;
+      }
+    } else {
+      r = driver->inject_lost_and_found(
+          obj_name_ino, file_size, file_mtime, chunk_size, data_pool_id);
+      if (r < 0) {
+        dout(4) << "Error injecting 0x" << std::hex << obj_name_ino
+          << std::dec << " into lost+found: " << cpp_strerror(r) << dendl;
+      }
+    }
+  }
+
+  return 0;
+}
+
+int MetadataDriver::read_fnode(inodeno_t ino, frag_t frag, fnode_t *fnode)
+{
+  assert(fnode != NULL);
+
+  object_t frag_oid = InodeStore::get_object_name(ino, frag, "");
+  bufferlist old_fnode_bl;
+  int r = metadata_io.omap_get_header(frag_oid.name, &old_fnode_bl);
+  if (r < 0) {
+    return r;
+  }
+
+  bufferlist::iterator old_fnode_iter = old_fnode_bl.begin();
+  try {
+    fnode_t old_fnode;
+    old_fnode.decode(old_fnode_iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+int MetadataDriver::read_dentry(inodeno_t parent_ino, frag_t frag,
+                const std::string &dname, InodeStore *inode)
+{
+  assert(inode != NULL);
+
+
+  std::string key;
+  dentry_key_t dn_key(CEPH_NOSNAP, dname.c_str());
+  dn_key.encode(key);
+
+  std::set<std::string> keys;
+  keys.insert(key);
+  std::map<std::string, bufferlist> vals;
+  object_t frag_oid = InodeStore::get_object_name(parent_ino, frag, "");
+  int r = metadata_io.omap_get_vals_by_keys(frag_oid.name, keys, &vals);  
+  dout(20) << "oid=" << frag_oid.name
+           << " dname=" << dname
+           << " frag=" << frag
+           << ", r=" << r << dendl;
+  if (r < 0) {
+    return r;
+  }
+
+  if (vals.find(key) == vals.end()) {
+    dout(20) << key << " not found in result" << dendl;
+    return -ENOENT;
+  }
+
+  try {
+    bufferlist::iterator q = vals[key].begin();
+    snapid_t dnfirst;
+    ::decode(dnfirst, q);
+    char dentry_type;
+    ::decode(dentry_type, q);
+    if (dentry_type == 'I') {
+      inode->decode_bare(q);
+      return 0;
+    } else {
+      dout(20) << "dentry type '" << dentry_type << "': cannot"
+                  "read an inode out of that" << dendl;
+      return -EINVAL;
+    }
+  } catch (const buffer::error &err) {
+    dout(20) << "encoding error in dentry 0x" << std::hex << parent_ino
+             << std::dec << "/" << dname << dendl;
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+int MetadataDriver::inject_lost_and_found(
+    inodeno_t ino, uint64_t file_size, time_t file_mtime, uint32_t chunk_size,
+    int64_t data_pool_id)
+{
+  // Create lost+found if doesn't exist
+  bool created = false;
+  int r = find_or_create_dirfrag(CEPH_INO_ROOT, frag_t(), &created);
+  if (r < 0) {
+    return r;
+  }
+  InodeStore lf_ino;
+  r = read_dentry(CEPH_INO_ROOT, frag_t(), "lost+found", &lf_ino);
+  if (r == -ENOENT || r == -EINVAL) {
+    // Inject dentry
+    lf_ino.inode.mode = 0755 | S_IFDIR;
+    lf_ino.inode.dirstat.nfiles = 1;
+    lf_ino.inode.size = 1;
```

Comment:
This is to fool code that skips doing readdirs for dirs with size 0 (adding a comment)